### PR TITLE
Fixes #36120 - Change default enablement for custom products

### PR DESCRIPTION
**Pull request by jeremylenz**

### CODE REVIEW

File: app/lib/actions/candlepin/product/content_add.rb [hunk 1]
```diff
@@ -10,7 +12,17 @@ class ContentAdd < Candlepin::Abstract
 
         def run
           output[:response] = ::Katello::Resources::Candlepin::Product.
-              add_content(input[:owner], input[:product_id], input[:content_id], true)
+              add_content(input[:owner], input[:product_id], input[:content_id], ENABLED)
```
Comment (jeremylenz) [inline app/lib/actions/candlepin/product/content_add.rb:15]: This is what sets it enabled in Candlepin (`enabled: 0` or `enabled: 1` will show in subscription-manager repos)

File: app/lib/actions/katello/product/content_create.rb [hunk 2]
```diff
@@ -52,9 +52,9 @@ def finalize
             new_content.vendor = ::Katello::Provider::CUSTOM
           end
 
-          #custom product content is always enabled by default
+          # custom product content is always disabled by default
           ::Katello::ProductContent.where(product: root.product, content: content).first_or_create do |pc|
-            pc.enabled = true
+            pc.enabled = false
```
Comment (jeremylenz) [inline app/lib/actions/katello/product/content_create.rb:57]: ..and this is what sets it in the Katello web UI / Repository Sets. We have to make sure we match Candlepin's value or bad things will happen.

**Comment by jeremylenz** [comment]
remember to test to make sure this doesn't break imports of RH repos!

File: app/lib/actions/candlepin/product/content_add.rb [hunk 3]
```diff
@@ -2,6 +2,8 @@ module Actions
   module Candlepin
     module Product
       class ContentAdd < Candlepin::Abstract
+        DEFAULT_ENABLEMENT = false
```
Comment (jeremylenz) [inline app/lib/actions/candlepin/product/content_add.rb:5]: When you're finished testing the PR, you can change this back to true and then run the rake task (and then remove all the content overrides ;)

**Comment by gvde** [comment]
> Currently this adds content overrides only for hosts. But I think it would be pretty easy to add them for activation keys too. Thoughts?

IMHO, activation keys are just as important and the migration should cover all existing hosts and all existing activation keys.

**Comment by jeremylenz** [comment]
@gvde Added activation keys to the rake task. :+1:

**Comment by gvde** [comment]
What about custom repositories in subscriptions which the host doesn't have? Those custom repositories should be overridden to disabled as well even though they are now enabled by default. Those would become visible when switching to SCA.

**Comment by jeremylenz** [comment]
> What about custom repositories in subscriptions which the host doesn't have? Those custom repositories should be overridden to disabled as well even though they are now enabled by default. Those would become visible when switching to SCA.

I'll expand more on community.theforeman.org, but after this upgrade, custom products will be disabled by default regardless of subscription status. So I don't think this should be an issue.

File: app/models/katello/product_content.rb [hunk 4]
```diff
@@ -3,6 +3,8 @@ class ProductContent < Katello::Model
     belongs_to :product, :class_name => 'Katello::Product', :inverse_of => :product_contents
     belongs_to :content, :class_name => 'Katello::Content', :inverse_of => :product_contents
 
+    validates :enabled, inclusion: { in: [true, false], message: N_("must be true or false") }
```
Comment (jeremylenz) [inline app/models/katello/product_content.rb:6]: Added this because what if `set_enabled_from_candlepin!` comes back with a `nil`..

**Comment by jeremylenz** [comment]
Updated the rake task to add disabled overrides for hosts and AKs in non-SCA orgs.

**Comment by jeremylenz** [comment]
updated the [description](https://github.com/Katello/katello/pull/10502#issue-1649958457)

**Comment by jeremylenz** [comment]
:green_apple:

**Comment by parthaa** [comment]
For the Rake task I would recommend using the approach here => https://github.com/Katello/katello/blob/master/app/lib/katello/util/deduplication_migrator.rb + https://github.com/Katello/katello/blob/master/db/migrate/20211201154845_add_unique_indexes.rb#L3 . That way you can actually unit test it like https://github.com/Katello/katello/blob/master/test/lib/util/deduplication_migrator_test.rb

**Comment by jeremylenz** [comment]
@parthaa Thanks, updated with new class and some tests :+1:

File: app/lib/katello/util/default_enablement_migrator.rb [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+              match_subscription: false,
+              match_environment: false,
+              consumable: host.subscription_facet
+            )
+            subscribed_content_finder = ::Katello::ProductContentFinder.new(
+              match_subscription: true,
+              match_environment: false,
+              consumable: host.subscription_facet
+            )
+            unsubscribed_content = content_finder.custom_content_labels - subscribed_content_finder.custom_content_labels
```
Comment (parthaa) [inline app/lib/katello/util/default_enablement_migrator.rb:43]: Can you explain the logic here. I guess if both `match_subscription` and `match_environment` is false do we get all the  product contents for an org?

File: app/lib/katello/util/default_enablement_migrator.rb [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+              match_subscription: false,
+              match_environment: false,
+              consumable: ak
+            )
+            subscribed_content_finder = ::Katello::ProductContentFinder.new(
+              match_subscription: true,
+              match_environment: false,
+              consumable: ak
+            )
+            unsubscribed_content = content_finder.custom_content_labels - subscribed_content_finder.custom_content_labels
```
Comment (parthaa) [inline app/lib/katello/util/default_enablement_migrator.rb:83]: Duplicated logic with the hosts and aks. Can we DRY this up?

**Comment by parthaa** [comment]
Something like
```patch
diff --git a/app/lib/katello/util/default_enablement_migrator.rb b/app/lib/katello/util/default_enablement_migrator.rb
index 652ff81b3d..8ed6657e93 100644
--- a/app/lib/katello/util/default_enablement_migrator.rb
+++ b/app/lib/katello/util/default_enablement_migrator.rb
@@ -22,6 +22,38 @@ module Katello
         Rails.logger.info("#{pluralize(kt_errors, "error")} updating default enablement in Katello; see log messages above") if kt_errors > 0
       end
 
+      def create_disabled_overrides_for_non_sca(consumable:)
+        content_finder = ::Katello::ProductContentFinder.new(
+                match_subscription: false,
+                match_environment: false,
+                consumable: consumable
+              )
+        subscribed_content_finder = ::Katello::ProductContentFinder.new(
+          match_subscription: true,
+          match_environment: false,
+          consumable: consumable
+        )
+        unsubscribed_content = content_finder.custom_content_labels - subscribed_content_finder.custom_content_labels
+        new_overrides = unsubscribed_content.map do |repo_label|
+          ::Katello::ContentOverride.new(
+            repo_label,
+            { name: "enabled", value: "0" } # Override to disabled
+          )
+        end
+        return if new_overrides.blank?
+        if consumable.is_a? ::Katello::Host::SubscriptionFacet
+          ::Katello::Resources::Candlepin::Consumer.update_content_overrides(
+            consumable.uuid,
+            new_overrides.map(&:to_entitlement_hash)
+          )
+        else
+          ::Katello::Resources::Candlepin::ActivationKey.update_content_overrides(
+            consumable.cp_id,
+            new_overrides.map(&:to_entitlement_hash)
+          )
diff --git a/app/lib/katello/util/default_enablement_migrator.rb b/app/lib/katello/util/default_enablement_migrator.rb
index 652ff81b3d..8ed6657e93 100644
--- a/app/lib/katello/util/default_enablement_migrator.rb
+++ b/app/lib/katello/util/default_enablement_migrator.rb
@@ -22,6 +22,38 @@ module Katello
         Rails.logger.info("#{pluralize(kt_errors, "error")} updating default enablement in Katello; see log messages above") if kt_errors > 0
       end
 
+      def create_disabled_overrides_for_non_sca(consumable:)
+        content_finder = ::Katello::ProductContentFinder.new(
+                match_subscription: false,
+                match_environment: false,
+                consumable: consumable
+              )
+        subscribed_content_finder = ::Katello::ProductContentFinder.new(
+          match_subscription: true,
+          match_environment: false,
+          consumable: consumable
+        )
+        unsubscribed_content = content_finder.custom_content_labels - subscribed_content_finder.custom_content_labels
+        new_overrides = unsubscribed_content.map do |repo_label|
+          ::Katello::ContentOverride.new(
+            repo_label,
+            { name: "enabled", value: "0" } # Override to disabled
+          )
+        end
+        return if new_overrides.blank?
+        if consumable.is_a? ::Katello::Host::SubscriptionFacet
+          ::Katello::Resources::Candlepin::Consumer.update_content_overrides(
+            consumable.uuid,
+            new_overrides.map(&:to_entitlement_hash)
+          )
+        else
+          ::Katello::Resources::Candlepin::ActivationKey.update_content_overrides(
+            consumable.cp_id,
+            new_overrides.map(&:to_entitlement_hash)
+          )
+        end
+      end
+
       # rubocop:disable Metrics/MethodLength
       def create_disabled_overrides_for_non_sca_org_hosts
         errors = 0
@@ -30,28 +62,7 @@ module Katello
           Rails.logger.info("Creating disabled overrides for unsubscribed content in org #{org.name}")
           hosts_to_update = org.hosts.where.not(subscription_facet: nil)
           hosts_to_update.each do |host|
-            content_finder = ::Katello::ProductContentFinder.new(
-              match_subscription: false,
-              match_environment: false,
-              consumable: host.subscription_facet
-            )
-            subscribed_content_finder = ::Katello::ProductContentFinder.new(
-              match_subscription: true,
-              match_environment: false,
-              consumable: host.subscription_facet
-            )
-            unsubscribed_content = content_finder.custom_content_labels - subscribed_content_finder.custom_content_labels
-            new_overrides = unsubscribed_content.map do |repo_label|
-              ::Katello::ContentOverride.new(
-                repo_label,
-                { name: "enabled", value: "0" } # Override to disabled
-              )
-            end
-            next if new_overrides.blank?
-            ::Katello::Resources::Candlepin::Consumer.update_content_overrides(
-              host.subscription_facet.uuid,
-              new_overrides.map(&:to_entitlement_hash)
-            )
+            create_disabled_overrides_for_non_sca(consumable: host.subscription_facet)
           rescue => e
             errors += 1
             Rails.logger.info("Failed to update host #{host.name}: #{e.message}")
@@ -70,28 +81,7 @@ module Katello
           Rails.logger.info("Creating disabled overrides for unsubscribed content in activation keys in org #{org.name}")
           aks_to_update = org.activation_keys
           aks_to_update.each do |ak|
-            content_finder = ::Katello::ProductContentFinder.new(
-              match_subscription: false,
-              match_environment: false,
-              consumable: ak
-            )
-            subscribed_content_finder = ::Katello::ProductContentFinder.new(
-              match_subscription: true,
-              match_environment: false,
-              consumable: ak
-            )
-            unsubscribed_content = content_finder.custom_content_labels - subscribed_content_finder.custom_content_labels
-            new_overrides = unsubscribed_content.map do |repo_label|
-              ::Katello::ContentOverride.new(
-                repo_label,
-                { name: "enabled", value: "0" } # Override to disabled
-              )
-            end
-            next if new_overrides.blank?
-            ::Katello::Resources::Candlepin::ActivationKey.update_content_overrides(
-              ak.cp_id,
-              new_overrides.map(&:to_entitlement_hash)
-            )
+            create_disabled_overrides_for_non_sca(consumable: ak)
           rescue => e
             errors += 1
             Rails.logger.info("Failed to update activation key #{activation key.name}: #{e.message}")
@@ -103,6 +93,31 @@ module Katello
         errors
       end
 
+      def create_overrides(consumable_id:, overriding_hosts: true)
+        repos_with_existing_overrides = ::Katello::Resources::Candlepin::ActivationKey.content_overrides(consumable_id).map do |override|
+          override[:contentLabel]
+        end
+        new_overrides = (repos_to_update - repos_with_existing_overrides).map do |repo_label|
+          ::Katello::ContentOverride.new(
+            repo_label,
+            { name: "enabled", value: "1" } # Override to enabled
+          )
+        end
+        return if new_overrides.blank?
+
+        if overriding_hosts
+          ::Katello::Resources::Candlepin::Consumer.update_content_overrides(
+            consumable_id,
+            new_overrides.map(&:to_entitlement_hash)
+          )
+        else
+          ::Katello::Resources::Candlepin::ActivationKey.update_content_overrides(
+            consumable_id,
+            new_overrides.map(&:to_entitlement_hash)
+          )
+        end
+      end
+
       def create_activation_key_overrides
         Rails.logger.info "Creating content overrides for all activation keys"
         ak_errors = 0
@@ -113,20 +128,7 @@ module Katello
           org_aks_to_update = ::Katello::ActivationKey.where(organization: org, cp_id: cp_aks_to_update)
           org_aks_to_update.each do |ak|
             Rails.logger.info "Updating activation key #{ak.name}"
-            repos_with_existing_overrides = ::Katello::Resources::Candlepin::ActivationKey.content_overrides(ak.cp_id).map do |override|
-              override[:contentLabel]
-            end
-            new_overrides = (repos_to_update - repos_with_existing_overrides).map do |repo_label|
-              ::Katello::ContentOverride.new(
-                repo_label,
-                { name: "enabled", value: "1" } # Override to enabled
-              )
-            end
-            next if new_overrides.blank?
-            ::Katello::Resources::Candlepin::ActivationKey.update_content_overrides(
-              ak.cp_id,
-              new_overrides.map(&:to_entitlement_hash)
-            )
+            create_overrides(consumable_id: ak.cp_id, overriding_hosts: false)
           rescue => e
             ak_errors += 1
             Rails.logger.info("Failed to update activation key #{ak.name}: #{e.message}")
@@ -148,22 +150,7 @@ module Katello
 
         consumers_to_update.each do |consumer_uuid|
           Rails.logger.info "Updating consumer #{consumer_uuid}"
-
-          # don't overwrite existing overrides
-          repos_with_existing_overrides = ::Katello::Resources::Candlepin::Consumer.content_overrides(consumer_uuid).map do |override|
-            override[:contentLabel]
-          end
-          new_overrides = (repos_to_update - repos_with_existing_overrides).map do |repo_label|
-            ::Katello::ContentOverride.new(
-              repo_label,
-              { name: "enabled", value: "1" } # Override to enabled
-            )
-          end
-          next if new_overrides.blank?
-          ::Katello::Resources::Candlepin::Consumer.update_content_overrides(
-            consumer_uuid,
-            new_overrides.map(&:to_entitlement_hash)
-          )
+          create_overrides(consumable_id: consumer_uuid, overriding_hosts: true)
         rescue => e
           consumer_errors += 1
           Rails.logger.info("Failed to update consumer #{consumer_uuid}: #{e.message}")

```

**Comment by jeremylenz** [comment]
@parthaa Updated :+1:

**Comment by chris1984** [comment]
@parthaa any other comments, will wait to test until you are done reviewing.

**Comment by parthaa** [comment]
Code wise LGTM. @chris1984 go ahead

**Review by chris1984 — APPROVED** [review]
ACK, nice work @jeremylenz very robust script.

The rake task does not log any errors - Pass ✅ 
```
2023-04-11T10:32:15 [I|app|] No change in enabled value for Candlepin content 1681223099730, ProductContent 8055
2023-04-11T10:32:15 [I|app|] Finished updating custom products enablement; no errors
```

### Host and activation key before:
scahost_before
host_before
ak_before
sca_ak_before

### After:

ak_after
sca_host_after
scaak_after


* Enabled overrides have been added to all custom products - Pass ✅ 
* Your existing overrides were not touched - Pass ✅ 
* Also verify the same for your activation keys - Pass ✅ 
* Check activation keys in multiple orgs - Pass ✅ 
* When you remove an override ("Reset to default") from a custom product, it now shows as "Disabled" and not "Enabled" - Pass ✅ 
host_step6
* When you create a new custom product with no overrides, it shows as "Disabled" - Pass ✅ 
host_step7
* For your non-SCA org, on both hosts and AKs, you now have disabled overrides for any repo that didn't have a subscription, and enabled overrides for any repo that does have a subscription. Note: Unless you turn on SCA, you'll only be able to see the non-subscription overrides on the Legacy Content Host UI, after checking the "Show All" checkbox. - Pass ✅ 
* IMPORTANT: Output of subscription-manager repos on the host always matches the UI: Disabled or Disabled (Overridden) in the UI = Enabled: 0; Enabled or Enabled (Overridden) in the UI = Enabled: 1 - Pass ✅ 
```
[root@rhel8 ~]# subscription-manager repos
+----------------------------------------------------------+
    Available Repositories in /etc/yum.repos.d/redhat.repo
+----------------------------------------------------------+
Repo ID:   Default_Organization_new-custom-product_new-custom-repo
Repo Name: new-custom-repo
Repo URL:  https://devel.croberts.io/pulp/content/Default_Organization/Library/custom/new-custom-product/new-custom-repo
Enabled:   0

Repo ID:   rhel-8-for-x86_64-baseos-rpms
Repo Name: Red Hat Enterprise Linux 8 for x86_64 - BaseOS (RPMs)
Repo URL:  https://devel.croberts.io/pulp/content/Default_Organization/Library/content/dist/rhel8/$releasever/x86_64/baseos/os
Enabled:   1

Repo ID:   rhel-8-for-x86_64-appstream-rpms
Repo Name: Red Hat Enterprise Linux 8 for x86_64 - AppStream (RPMs)
Repo URL:  https://devel.croberts.io/pulp/content/Default_Organization/Library/content/dist/rhel8/$releasever/x86_64/appstream/os
Enabled:   1

Repo ID:   ansible-2-for-rhel-8-x86_64-rpms
Repo Name: Red Hat Ansible Engine 2 for RHEL 8 x86_64 (RPMs)
Repo URL:  https://devel.croberts.io/pulp/content/Default_Organization/Library/content/dist/layered/rhel8/x86_64/ansible/2/os
Enabled:   0

Repo ID:   Default_Organization_Custom_Zoo
Repo Name: Zoo
Repo URL:  https://devel.croberts.io/pulp/content/Default_Organization/Library/custom/Custom/Zoo
Enabled:   1

Repo ID:   satellite-client-6-for-rhel-8-x86_64-rpms
Repo Name: Red Hat Satellite Client 6 for RHEL 8 x86_64 (RPMs)
Repo URL:  https://devel.croberts.io/pulp/content/Default_Organization/Library/content/dist/layered/rhel8/x86_64/sat-client/6/os
Enabled:   1
```

### Edge cases tested:
These were a pain in the butt to test because how fast the script runs :P

Tried turning on/off SCA on an org as the script runs - Pass ✅ 
Deleted a host while script runs - Pass ✅ 
Deleted an ak while script runs - Pass ✅ 
Non SCA ak, update subscription while script runs - Pass ✅ 
Trying to override a custom product while the script runs - Pass ✅ 
Refresh a manifest while script runs on both a small 1 sub manifest and a 30 sub employee long time refresh - Pass ✅
